### DISCUSSION TITLE
Trouble with bringup

### DISCUSSION THREAD

**Discussion by timmarkhuff** [Robotic-Decision-Making-Lab/blue#141] [Q&A]
@evan-palmer 
We have successfully installed the docker container and are trying to run the package now. So far we have been unable to make a connection. We are guessing a little bit on the steps. Maybe you can correct what we are doing:
1. connect to the Blue ROV 2 with QGroundControl. We are able to control the ROV and receive video feed.
2. run the container `docker run -it 6fdbd129a321`. This launches successfully.
3. enter the container `docker exec -it compose-blue-1 /bin/bash`. No problem here. 
4. run `ros2 launch blue_bringup bluerov2.launch.py` (we don't use the heavy configuration). This launches, but never connects. We see the following: 
```
[mavros_node-1] [INFO] [1697672363.943337332] [mavros]: Known MAVLink dialects: common ardupilotmega ASLUAV AVSSUAS all csAirLink cubepilot development icarous matrixpilot paparazzi standard storm32 uAvionix ualberta
[mavros_node-1] [INFO] [1697672363.943342951] [mavros]: MAVROS UAS via /uas1 started. MY ID 255.191, TARGET ID 1.1
[blue_manager-3] [INFO] [1697672364.511463271] [blue_manager]: Successfully backed up the thruster parameters from the parameters file.
[mavros_node-1] [INFO] [1697672364.845910860] [mavros_router]: link[1002] detected remote address 255.191
[mavros_node-1] [WARN] [1697672365.869374739] [mavros.sys]: VER: broadcast request timeout, retries left 4
[mavros_node-1] [WARN] [1697672366.868999505] [mavros.sys]: VER: broadcast request timeout, retries left 3
[mavros_node-1] [WARN] [1697672367.871337927] [mavros.sys]: VER: unicast request timeout, retries left 2
[mavros_node-1] [WARN] [1697672368.869431763] [mavros.sys]: VER: unicast request timeout, retries left 1
[mavros_node-1] [WARN] [1697672369.869823992] [mavros.sys]: VER: unicast request timeout, retries left 0
```
6. We have also tried running `docker run -it --net=host 6fdbd129a321`, thinking that might be necessary for the docker container to communicate with the Blue ROV 2, but that doesn't seem to help. 

Finally, we aren't sure what we should do after bring up. For now, we'd just like to do `ros2 topic echo` to confirm we are getting some data from the robot. Beyond that, what can we do? Any teleop or navigation launch files we should be running?

Thank you for your help.

**Comment by evan-palmer** [comment]
Great! So the issue that you are running into is that the default [MAVROS connection string](https://github.com/Robotic-Decision-Making-Lab/blue/blob/8abca640fd4540cbfbe4a045506fd04f1e8a0829/blue_description/config/mavros.yaml#L18) is setup for simulation. Are you running the container on a topside computer or from the BlueROV itself (you can do either)?

**Reply by timmarkhuff** [reply]
I am running the container on a topside computer, not trying to do simulation. Thanks for the tip. I will adjust the argument. What other things should we try once we bring it up? Is there a navigation node we can launch?

**Reply by evan-palmer** [reply]
Here is the connection string that I have used successfully in the past for topside operation:

```yaml
fcu_url: "udp://:14550@127.0.0.1:14549"
gcs_url: "udp://:14549@127.0.0.1:14548"
```
When you get things connected, I recommend testing out the MAVROS interface to make sure that you can set ArduSub flight modes, disarm/arm, etc. You may also want to adjust the plugins that MAVROS loads. There are quite a few that get loaded right now, and you may not need all of them. With regards to a navigation interface, I don't have Nav2 integrated yet, but it is on my [TODO list](https://github.com/Robotic-Decision-Making-Lab/blue/issues/35). As a heads up, I have run into a significant number of issues when using QGC alongside MAVROS (both with the BlueROV and drones). Once you get more familiar with things, I recommend not using QGC and instead using the joystick teleop interface and the camera interface for visualization in RViz

**Reply by evan-palmer** [reply]
If you end up running this on the BlueROV itself in the future, let me know and I'll put together some information regarding how to do that

**Reply by timmarkhuff** [reply]
Thank you! We will likely want to run some ROS nodes onboard the ROV in order to process images for localization, perhaps upgrading the RPI 3 to a RPI 5. We'll see how things progress...

**Comment by timmarkhuff** [comment]
Okay, we tried running ` ros2 launch blue_bringup bluerov2.launch.py use_sim:=false`, but it still isn't working. We see: 

```
[mavros_node-1] [INFO] [1697674279.830165396] [mavros]: Plugin trajectory ignored
[mavros_node-1] [INFO] [1697674279.830170967] [mavros]: Plugin tunnel ignored
[mavros_node-1] [INFO] [1697674279.830176526] [mavros]: Plugin vfr_hud ignored
[mavros_node-1] [INFO] [1697674279.830181896] [mavros]: Plugin vibration ignored
[mavros_node-1] [INFO] [1697674279.872691673] [mavros]: Plugin vision_pose created
[mavros_node-1] [INFO] [1697674279.872787004] [mavros]: Plugin vision_pose initialized
[mavros_node-1] [INFO] [1697674279.914068435] [mavros]: Plugin vision_speed created
[mavros_node-1] [INFO] [1697674279.914187903] [mavros]: Plugin vision_speed initialized
[mavros_node-1] [INFO] [1697674279.914207672] [mavros]: Plugin waypoint ignored
[mavros_node-1] [INFO] [1697674279.914214525] [mavros]: Plugin wheel_odomotry ignored
[mavros_node-1] [INFO] [1697674279.914219873] [mavros]: Plugin wind_estimation ignored
[mavros_node-1] [INFO] [1697674279.918169369] [mavros]: Built-in SIMD instructions: SSE, SSE2
[mavros_node-1] [INFO] [1697674279.918212269] [mavros]: Built-in MAVLink package version: 2023.9.9
[mavros_node-1] [INFO] [1697674279.918218866] [mavros]: Known MAVLink dialects: common ardupilotmega ASLUAV AVSSUAS all csAirLink cubepilot development icarous matrixpilot paparazzi standard storm32 uAvionix ualberta
[mavros_node-1] [INFO] [1697674279.918224263] [mavros]: MAVROS UAS via /uas1 started. MY ID 255.191, TARGET ID 1.1
[mavros_node-1] [INFO] [1697674280.807504929] [mavros_router]: link[1002] detected remote address 255.191
[blue_manager-3] [INFO] [1697674280.991824390] [blue_manager]: Successfully backed up the thruster parameters from the parameters file.
[mavros_node-1] [WARN] [1697674281.838231708] [mavros.sys]: VER: broadcast request timeout, retries left 4
[mavros_node-1] [WARN] [1697674282.837836017] [mavros.sys]: VER: broadcast request timeout, retries left 3
[mavros_node-1] [WARN] [1697674283.834394301] [mavros.sys]: VER: unicast request timeout, retries left 2
[mavros_node-1] [WARN] [1697674284.836373767] [mavros.sys]: VER: unicast request timeout, retries left 1
[mavros_node-1] [WARN] [1697674285.834356232] [mavros.sys]: VER: unicast request timeout, retries left 0
[mavros_node-1] [WARN] [1697674286.834169050] [mavros.sys]: VER: your FCU don't support AUTOPILOT_VERSION, switched to default capabilities
[mavros_node-1] [INFO] [1697674309.115739343] [mavros_router]: link[1000] trying to reconnect...
```

Any idea what is wrong? Did we miss a step?

**Reply by timmarkhuff** [reply]
Just saw your message on the previous thread. Let me give that a shot...

**Comment by Manouselis** [comment]
Hey @evan-palmer, maybe we can add to the documentation the three distinct config settings, as they're not intuitive:
The `ros_parameters` of the `mavros_node` that resides in `blue/blue_description/config/ardusub` should be changed from (**simulation**):
```
    fcu_url: "tcp://localhost"
    gcs_url: "udp://@localhost:14550"
```
to (**physical testing with docker image running locally ON THE ROVS CPU**):
```
    fcu_url: "udp://:14755@0.0.0.0:14755"
    gcs_url: "udp://@192.168.2.1:14550"
```
to (**physical testing with docker image running on the TOPSIDE computer**):
```
    fcu_url: "udp://:14550@127.0.0.1:14549"
    gcs_url: "udp://:14549@127.0.0.1:14548"
```

**Reply by evan-palmer** [reply]
Hey @Manouselis, I think this is a good idea. I have had it on my backlog for a while, but it's probably time to follow through with that.
